instance/event.go

```diff
@@ -28,37 +28,43 @@ func unmarshalEvent(reader io.Reader) (*models.Event, error) {
 	return &event, nil
 }
 
-//AddEvent details to an instance
+// AddEvent details to an instance
 func (s *Store) AddEvent(w http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
 	defer r.Body.Close()
 
 	vars := mux.Vars(r)
 	instanceID := vars["instance_id"]
+	eTag := getIfMatch(r)
 	data := log.Data{"instance_id": instanceID, "action": AddInstanceEventAction}
 
-	if err := func() error {
-		event, err := unmarshalEvent(r.Body)
-		if err != nil {
-			log.Event(ctx, "add instance event: failed to unmarshal request body", log.ERROR, log.Error(err), data)
-			return err
-		}
+	event, err := unmarshalEvent(r.Body)
+	if err != nil {
+		log.Event(ctx, "add instance event: failed to unmarshal request body", log.ERROR, log.Error(err), data)
+		handleInstanceErr(ctx, err, w, data)
+		return
+	}
 
-		if err = event.Validate(); err != nil {
-			log.Event(ctx, "add instance event: failed to validate event object", log.ERROR, log.Error(err), data)
-			return err
-		}
+	if err = event.Validate(); err != nil {
+		log.Event(ctx, "add instance event: failed to validate event object", log.ERROR, log.Error(err), data)
+		handleInstanceErr(ctx, err, w, data)
+		return
+	}
 
-		if err = s.AddEventToInstance(instanceID, event); err != nil {
-			log.Event(ctx, "add instance event: failed to add event to instance in datastore", log.ERROR, log.Error(err), data)
-			return err
-		}
+	instance, err := s.GetInstance(instanceID, eTag)
```

Comment:
Does this need wrapping with :
AcquireInstanceLock()
etc as done in dimension/dimension.go  ... about line 45'ish ?

Reply:
In this case, if the instance changes between GetInstance and AddEventToInstance, AddEventToInstance will fail (the mongoDB query filters by eTag, so if the eTag has changed, it will fail) - for dimensions/dimension.go, it would be possible that the instance changed before GetDimensionsFromInstance is called (as it checks a different collection, it's impossible to do it atomically unless we use a lock) - I thought in the event case it's not necessary and it would reduce the interactions with the database and possible delays in other calls that need the lock.

Reply:
Update: due to other calls requiring the lock, I will need to also acquire it here, even if this call itself wouldn't need it, to prevent concurrency issues with them.